packages/contracts/contracts/L1/ValidatorManager.sol

```diff
@@ -0,0 +1,497 @@
+    }
+
+    /**
+     * @notice Semantic version.
+     * @custom:semver 1.0.0
+     */
+    string public constant version = "1.0.0";
+
+    /**
+     * @notice Constructs the ValidatorManager contract.
+     *
+     * @param _constructorParams              Constructor parameters for AssetManager.
+     * @param _trustedValidator               Address of the trusted validator.
+     * @param _commissionRateMinChangeSeconds The minimum duration to change the commission rate in
+     *                                        seconds.
+     * @param _roundDurationSeconds           The duration of one submission round in seconds.
+     * @param _jailPeriodSeconds              The minimum duration to get out of jail in seconds.
+     * @param _jailThreshold                  The maximum allowed number of output non-submissions
+     *                                        before jailed.
+     */
+    constructor(
+        ConstructorParams memory _constructorParams,
+        address _trustedValidator,
+        uint128 _commissionRateMinChangeSeconds,
+        uint128 _roundDurationSeconds,
+        uint128 _jailPeriodSeconds,
+        uint128 _jailThreshold
+    ) AssetManager(_constructorParams) {
+        TRUSTED_VALIDATOR = _trustedValidator;
+        COMMISSION_RATE_MIN_CHANGE_SECONDS = _commissionRateMinChangeSeconds;
+
+        // Note that this value MUST be (SUBMISSION_INTERVAL * L2_BLOCK_TIME) / 2.
+        ROUND_DURATION_SECONDS = _roundDurationSeconds;
+
+        JAIL_PERIOD_SECONDS = _jailPeriodSeconds;
+        JAIL_THRESHOLD = _jailThreshold;
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function registerValidator(
+        uint128 assets,
+        uint8 commissionRate,
+        uint8 commissionMaxChangeRate
+    ) external {
+        require(
+            getStatus(msg.sender) == ValidatorStatus.NONE,
+            "ValidatorManager: already initiated validator"
+        );
+
+        require(
+            assets >= MIN_REGISTER_AMOUNT,
+            "ValidatorManager: need to register with at least min register amount"
+        );
+
+        require(
+            commissionRate <= COMMISSION_RATE_DENOM,
+            "ValidatorManager: the max value of commission rate has been exceeded"
+        );
+
+        require(
+            commissionMaxChangeRate <= COMMISSION_RATE_DENOM,
+            "ValidatorManager: the max value of commission rate max change rate has been exceeded"
+        );
+
+        Vault storage vault = _vaults[msg.sender];
+        vault.isInitiated = true;
+        vault.reward.commissionRate = commissionRate;
+        vault.reward.commissionMaxChangeRate = commissionMaxChangeRate;
+        vault.reward.commissionRateChangedAt = uint128(block.timestamp);
+
+        _delegate(msg.sender, msg.sender, assets, false);
+
+        bool canStart = assets >= MIN_START_AMOUNT;
+        if (canStart) {
+            _validatorTree.insert(msg.sender, uint120(_reflectiveWeight(vault)));
+        }
```

Comment:
How about this?

```suggestion
        bool canStart = assets >= MIN_START_AMOUNT;

        if (canStart) {
            _delegate(msg.sender, msg.sender, assets, true);
        } else {
            _delegate(msg.sender, msg.sender, assets, false);
        }
```

Reply:
Because in `_delegate`, we do `tree.update`, but here we need to do `tree.insert`.

Reply:
Got it, I missed that point.